core/graph/src/main/java/uk/gov/gchq/gaffer/graph/Graph.java

```diff
@@ -285,7 +285,7 @@ public JobDetail executeJob(final Job job, final Context context) throws Operati
                 try {
                     result = graphHook.onFailure(result, clonedOpChain, clonedContext, e);
                 } catch (final Exception graphHookE) {
-                    LOGGER.warn("Error in graphHook " + graphHook.getClass().getSimpleName() + ": " + graphHookE.getMessage(), graphHookE);
+                    LOGGER.warn(String.format("Error in graphHook %s : %s", graphHook.getClass().getSimpleName(), graphHookE.getMessage()), graphHookE);
```

Comment:
@t92549 In the issue (#2618), you say that concatenation "is considered bad practice as the concatenated string object is always created, whether or not that string will be logged". Yet this solution has exactly the same problem. The `String.format` static call will always be made regardless of logging level.

SLF4J doesn't offer a variant which accepts a formatting string, objects to be added into the string **and** an exception/throwable. Although it could offer it, considering how the formatting is done (see [here](https://github.com/qos-ch/slf4j/blob/82a592ed790cc8072aef1455f1065b6c9d207501/slf4j-api/src/main/java/org/slf4j/helpers/AbstractLogger.java) & [here](https://github.com/qos-ch/slf4j/blob/82a592ed790cc8072aef1455f1065b6c9d207501/slf4j-reload4j/src/main/java/org/slf4j/reload4j/Reload4jLoggerAdapter.java#L133)).

In this example, `graphHookE` is the exception and I presume we want to log the stack trace (as this is what happens when we include it in the call to the logger). Why is it that we sometimes do this and not others? If we don't really need the stack trace then we shouldn't include it and thereby remove a need to create the string object.

Reply:
@GCHQDeveloper314 Is there anyway to work around the fact that we need to use String.format or + to concatenate the strings? Maybe make a variable with the msg value and assign it to the function?

Reply:
You are right, just did some testing and I apologise, indeed the best way to go about this is something like:
```java
LOGGER.warn("Error in graphHook {} : {}", graphHook.getClass().getSimpleName(), graphHookE.getMessage(), graphHookE); 
```

Reply:
There's not an easy way to work around it. [As is written in the Javadoc](https://github.com/qos-ch/slf4j/blob/82a592ed790cc8072aef1455f1065b6c9d207501/slf4j-api/src/main/java/org/slf4j/Logger.java#L613), the varargs variant still causes an object to be created (although this is still better than concatenation). It is better for readability, but I think it might be arguable that if logging is enabled it could be less efficient vs `String.format` considering the extra work required to determine if the final object is a thowable. Another different option would be to convert the exception to a string before passing it in ([could use this](https://commons.apache.org/proper/commons-lang/apidocs/org/apache/commons/lang3/exception/ExceptionUtils.html#getStackTrace-java.lang.Throwable-)), although that always creates an extra object which might not get logged and is probably less efficient.

The best (most efficient) solution would be to make a helper class (that extends SLF4J) which would take all the parameters we want and not do any unnecessary processing. That is far out of scope of this issue and most likely not something we'd want to do anyway. Ensuring we only log exceptions we need (to reduce overhead more generally) could be raised as another issue but it's pretty trivial really.

Ultimately, the above is mostly out of scope of this issue. As you have pointed out, there's no need for the String format and the `Exception` can be added as the final argument.

So far as this PR is concerned, `String.format` is not required and so shouldn't be used. With that done, we just need to remove the remaining instances of string concatenation. Using `grep -EIr "LOGGER.*\+"` I can see another 8 remaining in the codebase. Once that is all done this can be merged in.